secrecy/src/boxed.rs

```diff
@@ -1,10 +1,77 @@
 //! `Box` types containing secrets
 
-use super::{DebugSecret, Secret};
 use alloc::boxed::Box;
-use zeroize::Zeroize;
+use core::{
+    any,
+    fmt::{self, Debug},
+};
 
-/// `Box` types containing a secret value
-pub type SecretBox<S> = Secret<Box<S>>;
+use zeroize::{Zeroize, ZeroizeOnDrop};
 
-impl<S: DebugSecret + Zeroize> DebugSecret for Box<S> {}
+use crate::{ExposeSecret, ExposeSecretMut};
+
+/// Same as [`Secret`], but keeps the secret value in the heap instead of on the stack.
+pub struct SecretBox<S: Zeroize> {
+    inner_secret: Box<S>,
+}
+
+impl<S: Zeroize + Clone> SecretBox<S> {
+    /// Create a secret value using the provided function as a constructor.
+    ///
+    /// The implementation makes an effort to zeroize the locally constructed value
+    /// before it is copied to the heap, and constructing it inside the closure minimizes
+    /// the possibility of it being accidentally copied by other code.
+    pub fn new(ctr: impl FnOnce() -> S) -> Self {
+        let mut data = ctr();
+        let secret = Self {
+            inner_secret: Box::new(data.clone()),
+        };
+        data.zeroize();
+        secret
+    }
```

Comment:
My first thought looking at this is that there has to be another way and that this approach is suboptimal.

I am unfortunately quite aware of how tricky it is to ensure values are constructed on the heap, but this approach seems like it would make ensured inlining more difficult and force the value to be stack allocated, *then* moved to the heap, whereas with proper inlining I believe it's possible to avoid that initial stack allocation.

It's been awhile since I've checked that in godbolt though, and I think work on placement-by-return has largely stalled (and the `box` keyword removed).

Reply:
A simpler alternative is to accept `Box<T>` as an argument.

I think a complex constructor like this needs to be informed by and justified by the generated assembly. If it leaves any copies on the stack it defeats the point (and potentially the inliner).

Reply:
> I am unfortunately quite aware of how tricky it is to ensure values are constructed on the heap, but this approach seems like it would make ensured inlining more difficult and force the value to be stack allocated, then moved to the heap, whereas with proper inlining I believe it's possible to avoid that initial stack allocation.

I don't think it is supported in Rust currently. It's called "placement new", and currently only exists [as a proposal](https://github.com/PoignardAzur/placement-by-return/blob/placement-by-return/text/0000-placement-by-return.md). At the moment we have to accept the fact the we have to construct the value on stack, and all we can do is at least contain it within a small closure and zeroize it. 

> A simpler alternative is to accept Box<T> as an argument.

That is certainly a good idea as an additional constructor, but it's not always possible to already have a `Box`.

Reply:
> I don't think it is supported in Rust currently. It's called "placement new", and currently only exists [as a proposal](https://github.com/PoignardAzur/placement-by-return/blob/placement-by-return/text/0000-placement-by-return.md).

You just linked to the placement-by-return proposal I was referring to in my comments.

But without placement-by-return, you're going to make a copy when `ctr` returns a value, which defeats the whole point.

The only way to eliminate the copy is by eliminating the intermediate stack frame, i.e. inlining, which it seems like this approach will always defeat.

Reply:
In other words, this constructor needlessly forces a stack allocation of a value, *then* copies it to the heap.

I think we should avoid making copies as much as possible.

Reply:
> Not according to my tests.

What tests did you do?

Reply:
> In other words, this constructor needlessly forces a stack allocation of a value, then copies it to the heap.

Well, either you have a `Box::new(T::default())` and then mutate it, or you construct it on the stack. The former is not always possible or convenient, and the latter seems to be safe with the constructor I provided.

> What tests did you do?

Creating `Secret` values in a function and inspecting the stack with `psm` and `read-process-memory` I linked in the description. I'm thinking on how to make it into an autotest, since it's a little flaky.

Reply:
I think your proposed constructor would need a lot of testing on different platforms to determine it accomplishes the desired goals, yes. Can we hold off on it for now?

Reply:
The `Clone` bound is another argument: since the only constructor requires a `Clone` bound, it means all your secret types *MUST* impl `Clone`, which may be undesirable.

Reply:
We can if we must, but I'll probably still have to use it in my own code since I cannot create a random `crypto_bigint::Uint` or `k256::Scalar` directly in a `Box`.